From 426797f9c0723dd7d37aa6dd294f9ddbf798b2dd Mon Sep 17 00:00:00 2001
From: Yves Rutschle <git1@rutschle.net>
Date: Sun, 30 Mar 2014 17:28:00 +0200
Subject: [PATCH] call setgroups before setgid

	Call setgroups() before setgid() (fixes potential
	privilege escalation).
	(Lars Vogdt)

--- common.c
+++ common.c
@@ -6,6 +6,7 @@
 
 #define _GNU_SOURCE
 #include <stdarg.h>
+#include <grp.h>
 
 #include "common.h"
 #include "probe.h"
@@ -578,6 +579,13 @@ void drop_privileges(const char* user_name)
 
     set_keepcaps(1);
 
+    /* remove extraneous groups in case we belong to several extra groups that
+     * may have unwanted rights. If non-root when calling setgroups(), it
+     * fails, which is fine because... we have no unwanted rights 
+     * (see POS36-C for security context)
+     * */
+    setgroups(0, NULL);
+
     res = setgid(pw->pw_gid);
     CHECK_RES_DIE(res, "setgid");
     res = setuid(pw->pw_uid);
